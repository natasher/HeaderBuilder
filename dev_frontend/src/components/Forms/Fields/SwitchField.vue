<template>
  <div class = "row">

    <div class = "col col-left">
      <span class = "title">{{ fieldName }}</span>
    </div>

    <div class = "col col-right">

      <base-switch
        :fieldName    = "fieldName"
        :wpId         = "wpId"
        :label        = "label"
        :isActive     = "getCurrentFieldValue( wpId )"
        @click.native = "toggleSwitchState({ name: wpId })" />

    </div>

  </div>
</template>

<script>
import { mapGetters, mapActions } from 'vuex'
import BaseSwitch from '../../BaseGeneric/BaseSwitch.vue'

export default {

  name: 'SwitchField',

  components: {
    BaseSwitch,
  },

  props: {
    /**
     * The name of the field.
     * Displayed on the left column.
     */
    fieldName: {
      type    : String,
      required: true,
    },
    /**
     * Name of the field used by WordPress internals.
     */
    wpId: {
      type   : String,
      require: true,
    },
    /**
     * Label shown next to switch.
     */
    label: {
      type    : String,
      required: false,
    },
  },

  computed: {
    ...mapGetters( 'fields', [ 'getCurrentFieldValue', ]),
  },

  methods: {
    ...mapActions( 'fields', [ 'toggleSwitchState', ]),
  },

}
</script>
