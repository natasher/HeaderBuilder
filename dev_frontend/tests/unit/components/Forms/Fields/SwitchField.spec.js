import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex          from 'vuex'
// import flushPromises from 'flush-promises'
import merge         from 'lodash.merge'
import SwitchField   from '../../../../../src/components/Forms/Fields/SwitchField.vue'

const localVue = createLocalVue()
localVue.use( Vuex )

function createStore(overrides) {
  const defaultStoreConfig = {
    modules: {
      fields: {
        namespaced: true,
        actions: {
          toggleSwitchState: jest.fn()
        },
        getters: {
          getSwitchFieldValue: jest.fn().mockReturnValue(
            jest.fn().mockReturnValue( true )
          )
        }
      }
    }
  }

  return new Vuex.Store(
    merge( defaultStoreConfig, overrides )
  )
}

const propsData = {
  fieldName: 'Test Field',
  wpId     : 'testField',
  label    : 'test label',
}

function createWrapper(overrides) {
  const defaultMountingOptions = {
    localVue,
    store: createStore(),
    propsData,
  }

  return shallowMount( SwitchField, merge( defaultMountingOptions, overrides ))
}

describe('SwitchField.vue', () => {

  test('should match snapshot', () => {
    const wrapper = createWrapper()

    expect( wrapper.element ).toMatchSnapshot()
  })

  test('should toggle switch if user click switch', async () => {
    const toggleSwitchState = jest.fn()
    const store = createStore({
      modules: {
        fields: {
          actions: {
            toggleSwitchState,
          }
        }
      }
    })
    const wrapper = createWrapper({ store, })
    const sw      = wrapper.find('baseswitch-stub')
    sw.trigger( 'click' )

    expect(toggleSwitchState).toBeCalledWith(
      expect.anything(),
      {
        name: wrapper.vm.wpId,
        as  : wrapper.vm.as
      },
      void 0
    )
  })

})