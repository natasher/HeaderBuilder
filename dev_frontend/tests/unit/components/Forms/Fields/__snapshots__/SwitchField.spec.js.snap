// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwitchField.vue should match snapshot 1`] = `
<div
  class="row"
>
  <div
    class="col col-left"
  >
    <span
      class="title"
    >
      Test Field
    </span>
  </div>
   
  <div
    class="col col-right"
  >
    <baseswitch-stub
      fieldname="Test Field"
      isactive="true"
      label="test label"
      wpid="testField"
    />
  </div>
</div>
`;
